lxd/patches.go

```diff
@@ -141,3 +145,842 @@ func patchNetworkPermissions(name string, d *Daemon) error {
+	// independent of whether anything currently exists on that pool. We can
+	// still probably safely assume that the user at least once used that
+	// pool.
+	// However, when we detect {dir, btrfs}, we can't rely on that guess
+	// since the daemon doesn't record any name for the pool anywhere.  So
+	// in the {dir, btrfs} case we check whether anything exists on the
+	// pool, if not, then we don't create a default pool. The user will then
+	// be forced to run lxd init again and can start from a pristine state.
+	// COMMENT(brauner): Check if this LXD instace currently has any
+	// containers, snapshots, or images configured. If so, we create a
+	// default storage pool in the database. Otherwise, the user will have
+	// to run LXD init.
+	cRegular, err := dbContainersList(d.db, cTypeRegular)
+	if err != nil {
+		return err
+	}
+
+	// COMMENT(brauner): Get list of existing snapshots.
+	cSnapshots, err := dbContainersList(d.db, cTypeSnapshot)
+	if err != nil {
+		return err
+	}
+
+	// COMMENT(brauner): Get list of existing public images.
+	imgPublic, err := dbImagesGet(d.db, true)
+	if err != nil {
+		return err
+	}
+
+	// COMMENT(brauner): Get list of existing private images.
+	imgPrivate, err := dbImagesGet(d.db, false)
+	if err != nil {
+		return err
+	}
+
+	// COMMENT(brauner): Nothing exists on the pool so we're not creating a
+	// default one, thereby forcing the user to run lxd init.
+	if len(cRegular) == 0 && len(cSnapshots) == 0 && len(imgPublic) == 0 && len(imgPrivate) == 0 {
+		return nil
+	}
+
+	switch preStorageApiStorageType {
+	case storageTypeBtrfs:
+		err = upgradeFromStorageTypeBtrfs(name, d, defaultPoolName, defaultStorageTypeName, cRegular, cSnapshots, imgPublic, imgPrivate)
+	case storageTypeDir:
+		err = upgradeFromStorageTypeDir(name, d, defaultPoolName, defaultStorageTypeName, cRegular, cSnapshots, imgPublic, imgPrivate)
+	case storageTypeLvm:
+		err = upgradeFromStorageTypeLvm(name, d, defaultPoolName, defaultStorageTypeName, cRegular, cSnapshots, imgPublic, imgPrivate)
+	case storageTypeZfs:
+		err = upgradeFromStorageTypeZfs(name, d, defaultPoolName, defaultStorageTypeName, cRegular, []string{}, imgPublic, imgPrivate)
+	default: // Shouldn't happen.
+		return fmt.Errorf("Invalid storage type. Upgrading not possible.")
+	}
+	if err != nil {
+		return err
+	}
+
+	err = daemonConfig["storage.default_pool"].Set(d, defaultPoolName)
+	if err != nil {
+		return err
+	}
+
+	containers := append(cRegular, cSnapshots...)
+	for _, ct := range containers {
+		c, err := dbContainerGet(d.db, ct)
+		if err != nil {
+			shared.LogWarnf("Failed to add \"pool\" property to containers root device.")
+			continue
+		}
+
+		for k, v := range c.Devices {
+			if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" {
+				c.Devices[k]["pool"] = defaultPoolName
+
+				// COMMENT(brauner): Ugh, seems like the only
+				// way to update devices currently.
+				tx, err := dbBegin(d.db)
```

Comment:
Don't you want to do all of this in one big transaction, instead of one transaction for each device?